# Confidence intervals case study using NHANES data


This notebook demonstrates how to use Python and its statistical libraries to construct confidence intervals for proportions and means.  We will also cover some important points relating to the properties of confidence intervals, and discuss how to use and interpret confidence intervals in practice.  We will use the 2015-2016 wave of the [NHANES](https://www.cdc.gov/nchs/nhanes/index.htm) data for all the analyses below.

It is important to note that the NHANES data are a "complex survey".  The data are not an independent and representative sample from the target population.  Proper analysis of complex survey data should make use of additional information about the manner in which the data were collected.  Since complex survey analysis is a somewhat specialized topic, we ignore this aspect of the data here, and analyze the NHANES data as if it were an independent and identically distributed sample from a population.

The following module import statements and data reading statement are identical to what we have used previously in course 1.

In [1]:
%matplotlib inline
import pandas as pd
import numpy as np
import seaborn as sns
import statsmodels.api as sm
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv('data/nhanes_2015_2016.csv')
df.head(3)

SEQN  ALQ101  ALQ110  ALQ130  SMQ020  RIAGENDR  RIDAGEYR  RIDRETH1  \
0  83732     1.0     NaN     1.0       1         1        62         3   
1  83733     1.0     NaN     6.0       1         1        53         3   
2  83734     1.0     NaN     NaN       1         1        78         3   

   DMDCITZN  DMDEDUC2  ...  BPXSY2  BPXDI2  BMXWT  BMXHT  BMXBMI  BMXLEG  \
0       1.0       5.0  ...   124.0    64.0   94.8  184.5    27.8    43.3   
1       2.0       3.0  ...   140.0    88.0   90.4  171.4    30.8    38.0   
2       1.0       3.0  ...   132.0    44.0   83.4  170.1    28.8    35.6   

   BMXARML  BMXARMC  BMXWAIST  HIQ210  
0     43.6     35.9     101.1     2.0  
1     40.0     33.2     107.9     NaN  
2     37.0     31.0     116.5     2.0  

[3 rows x 28 columns]

## Confidence intervals for one proportion

In this section, we demonstrate the construction of confidence intervals for the proportion of people who smoke.  The specific definition of "smoker" used here ([SMQ020](https://wwwn.cdc.gov/Nchs/Nhanes/2015-2016/SMQ_I.htm#SMQ020)) identifies a person as being a smoker if they self-report as having smoked 100 or more cigarettes in their lifetime.  It is more accurate to refer to this as a measure of "lifetime smoking" rather than "current smoking".  Recall that the definitions of these and other NHANES variables can be found using the NHANES code books, or by searching using the link below.

https://wwwn.cdc.gov/nchs/nhanes/search/default.aspx

We will calculate the proportions of smokers separately for females and for males.  Initially we can compare these two proportions and their corresponding confidence intervals informally, but later we will discuss methods to compare two proportions formally using confidence intervals.

First we replace the numeric codes in the variables of interest with text labels, and set the rare answers other than "yes" and "no" to be missing (so they will automatically be omitted from all the analyses below).

In [3]:
df['SMQO20x'] = df['SMQ020'].replace({1: 'yes',
                                      2: 'no',
                                      7: np.nan,
                                      9: np.nan})
df['RIAGENDRx'] = df['RIAGENDR'].replace({1: 'male',
                                          2: 'female'})


We can now [tabulate](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.crosstab.html) the numbers of female and male smokers and non-smokers:

In [4]:
dx = df[['SMQO20x', 'RIAGENDRx']].dropna()
dx.head(3)

SMQO20x RIAGENDRx
0     yes      male
1     yes      male
2     yes      male

In [5]:
pd.crosstab(dx.SMQO20x, dx.RIAGENDRx)

RIAGENDRx  female  male
SMQO20x                
no           2066  1340
yes           906  1413

In [6]:
dx

SMQO20x RIAGENDRx
0        yes      male
1        yes      male
2        yes      male
3         no    female
4         no    female
...      ...       ...
5730     yes    female
5731      no      male
5732     yes    female
5733     yes      male
5734      no    female

[5725 rows x 2 columns]

The confidence interval (CI) is constructed using two inputs: the sample proportion of smokers, and the total sample size for smokers and non-smokers combined. We calculate these values next.

In [7]:
dx['SMQO20x'].replace({'yes': 1, 'no': 0}, inplace=True)
dx['RIAGENDRx'].replace({'male': 1, 'female': 0}, inplace=True)

dz = dx.groupby('RIAGENDRx').agg(['mean', 'count'])
dz.columns = ['Proportion', 'total']
dz

Proportion  total
RIAGENDRx                   
0            0.304845   2972
1            0.513258   2753

Confidence intervals are closely connected to standard errors. Recall that the standard error essentially tells you how far you should expect an estimate to fall from the truth. A confidence interval is an interval that under repeated sampling covers the truth a defined proportion of the time. In most settings, this "coverage probability" is set to 95%.

It turns out that in many settings, a 95% confidence interval can be constructed as the interval consisting of all points that are within two (or 1.96) standard errors of the point estimate. More concisely, the confidence interval approximately spans from e - 2•SE to e + 2•SE, where e is the point estimate and SE is the standard error.

Since the standard error plays such an important role here, we calculate it separately first.

In [8]:
dz.Proportion[0]

0.30484522207267833

In [9]:
# female
p = dz.Proportion[0]
n = dz.total[0]
se_female = np.sqrt(p * (1 - p) / n)

# male
p = dz.Proportion[1]
n = dz.total[1]
se_male = np.sqrt(p * (1 - p) / n)

se_female, se_male

(0.008444152146214435, 0.009526078653689868)

We can see that the standard errors for the estimated proportions of females and males who smoke are similar, and are each around 1% (since we are studying a proportion here, 0.01 corresponds to a 1 percentage point change in the smoking rate).

The standard error for a proportion is maximized when the true proportion is around 1/2, and gets smaller as the true proportion approaches either 0 or 1. The estimated male smoking proportion is closer to 1/2 than the estimated female smoking proportion, and the male sample size is smaller than the female sample size. Both of these factors lead to the male standard error being larger than the female standard error, although the difference is very small in this case.

Next we calculate the 95% confidence intervals for the proportions of female and male smokers using the formula for the one-sample confidence interval for a proportion:

In [10]:
# female
p = dz.Proportion[0]
n = dz.total[0]

moe = 1.96 * np.sqrt(p * (1 - p) / n)
lower_bound, upper_bound = p - moe, p + moe
lower_bound, upper_bound

(0.288294683866098, 0.32139576027925865)

The results above indicate that any population proportion (for female lifetime smokers) between 0.288 and 0.321 would be compatible with the data that we observed in NHANES.

In [11]:
# male
p = dz.Proportion[1]
n = dz.total[1]

moe = 1.96 * np.sqrt(p * (1 - p) / n)
lower_bound, upper_bound = p - moe, p + moe
lower_bound, upper_bound

(0.49458714955108174, 0.531929377873546)

These results indicate that any population proportion (for male lifetime smokers) between 0.493 and 0.531 would be compatible with the NHANES data.

In a routine data analysis, we do not need to calculate these intervals manually. We can use the Statsmodels library to calculate the CI for us in one line:

In [12]:
# male
print('Male:  ', sm.stats.proportion_confint(1413, 1413 + 1340))

# female
print('Female:', sm.stats.proportion_confint(906, 906 + 2066))

Male:   (0.49458749263718593, 0.5319290347874418)
Female: (0.2882949879861214, 0.32139545615923526)


The confidence intervals for the proportions of female and male smokers shown above are quite narrow and do not overlap.  This suggests that there is a substantial difference between the lifetime smoking rates for women and men. However there is no explicit information here about how different the two population proportions might be.  To address this question, we can form a confidence interval for the difference between the proportion of females who smoke and the proportion of males who smoke.

The point estimate of the difference between female and male smoking rates is -0.208 (0.305 - 0.513).  That is, the smoking rate is about 20 percentage points higher in men than in women.  This difference of around 20 percentage points is only a point estimate of the underlying true value -- it is not exactly equal to the difference between the unknown proportions of females and males who smoke in the population.  A confidence interval helps us assess how far the estimated difference may be from the true difference.

As above, we start with the standard error.  The difference between two sample proportions based on independent data has a standard error that reflects the combined uncertainty in the two proportions being differenced.  This standard error can be calculated very easily.  If SE1 and SE2 are the standard errors for two proportions, then sqrt(SE1^2 + SE2^2) is the standard error for the difference of these proportions (`sqrt` is the square root function).  Note that this formula is only accurate if the two sample proportions being differenced are independent.

In the next cell we calculate the standard error for the difference between the proprotion of females who smoke and the proportion of males who smoke.

In [13]:
se_diff = np.sqrt(se_female**2 + se_male**2)
se_diff

0.012729881381407434

The standard error of around 0.013 indicates that the estimated difference statistic -0.208 is expected to fall around 0.013 units from the true value. We do not know in which direction the error lies, and we do not know that the error is exactly 0.013, only that it is around this large on average. For most purposes, a standard error of 0.013 relative to an observed difference of -0.21 would be considered very small. That is, we have a very accurate estimate of the difference between smoking rates in women and in men.

Now that we have the standard error, we can construct a 95% confidence interval for the difference in proportions by taking the estimate and subtracting and adding two (or 1.96) standard errors from it.

In [14]:
d = dz.Proportion[0] - dz.Proportion[1]
lcb = d - 2 * se_diff
ucb = d + 2 * se_diff
print(lcb, ucb)

-0.2338728044024504 -0.18295327887682067


The 95% confidence interval above shows us that any value for the difference of population proportions (between females and males) lying between -0.233 and -0.183 is consistent with the observed data.

## Confidence intervals for sub-populations

Since smoking rates vary strongly with age, it might be more informative to stratify the data into homogeneous age bands and compare the proportions of female and male smokers within each age band. We can also calculate the 95% confidence interval for this difference within each age band. These data can be displayed as a plot, with the difference in proportions plotted as a curve. The confidence intervals can then be used to construct a "confidence band" around the estimates.


In [15]:
# calculate the smoking rates within age/gender groups
df["agegrp"] = pd.cut(df.RIDAGEYR, [18, 30, 40, 50, 60, 70, 80])
pr = df.groupby(["agegrp", "RIAGENDRx"]).agg({"SMQ020": lambda x: np.mean(x == 1)}).unstack()
pr.columns = ["female", "male"]
pr

female      male
agegrp                      
(18, 30]  0.226601  0.349265
(30, 40]  0.286920  0.502183
(40, 50]  0.268924  0.448878
(50, 60]  0.421277  0.572687
(60, 70]  0.374150  0.654462
(70, 80]  0.324390  0.649254

In [16]:
# The number of people for each calculated proportion
dn = df.groupby(['age_group', 'RIAGENDRx']).agg({'SMQ020': np.size}).unstack()
dn.columns = ['female', 'male']
dn

KeyError: 'age_group'

In [ ]:
# standard errors for each proportion
se = np.sqrt(pr * (1 - pr) / dn)

# standard error for the difference in female/male smoking rates in every age group
se_diff = np.sqrt(se.female ** 2 + se.male ** 2)
se_diff

In [ ]:
# Standard errors for the difference in smoking rates between genders, within age bands

# The difference in smoking rates between genders
pq = pr.female - pr.male

x = np.arange(pq.size)
pp = sns.pointplot(x, pq.values, color='black')
sns.pointplot(x, pq - 2 * se_diff)
sns.pointplot(x, pq + 2 * se_diff)
pp.set_xticklabels(pq.index)
pp.set_xlabel("Age group")
pp.set_ylabel("Female - male smoking proportion")

The plot above shows for each age band, the point estimate of the difference in smoking rates between genders (black dot), and the lower and upper end points of the 95% confidence interval (blue points). Based on this plot, we see that in the United States, smoking is more common in men than in women, not just overall, but also in every one of the age bands. The difference is largest for older people -- for people older than 60, the smoking rate for males is around 30 percentage points greater than the smoking rate for females, while for people younger than 30, the smoking rate for males is only around 15 percentage points greater than the smoking rate for females.

Also note that the 95% confidence bands shown above are much wider than the 95% confidence intervals for the data that were not stratified by age. Stratifying by age leads to smaller sample sizes, which in turn results in wider confidence intervals.

## Confidence intervals for the mean

In this section, we discuss how to construct confidence intervals for the mean.  First note that the proportion discussed above is also a mean -- for example, if the data are 0, 1, 0, then the mean is 1/3, which is also the proportion of 1's in the data.  However the proportion has the special property that the variance is completely determined by the mean. That is why we constructed the standard errors for the sample proportion above using p•(1 - p) as the variance.  In general, the variance of quantitative data will not be a function of the mean, as this is a very special property of binary data.  Therefore, in general we must estimate the variance as a separate step after estimating the mean.

To illustrate the construction of confidence intervals for the population mean of a quantitative variable, we will use the body mass inde (BMI) data from NHANES.  To begin, we calculate the mean BMI for all women and for all men in the NHANES sample.

In [ ]:
df.groupby('RIAGENDRx').agg({'BMXBMI': 'mean'})

The numbers in the first column of the table above are estimates of the population mean BMI for all women and for all men in the United States (the population that the NHANES study represents). As with the sample proportions, these numbers are not exactly equal to the mean BMI for all women and men, they are only estimates. To establish the uncertainty for these estimates, we can use the standard errors for these two estimated means.

The standard error for the mean based on an independent and identically distributed sample is equal to the standard deviation of the variable divided by the square root of the sample size. We next calculate all the relevant values needed to compute the standard error.

In [ ]:
df.groupby('RIAGENDRx').agg({'BMXBMI': ['mean', 'std', 'count']})

We can now calculate the standard error of the mean BMI for women and for men:

In [ ]:
se_female = 7.753319 / np.sqrt(2944)
se_male   = 6.252568 / np.sqrt(2718)

se_female, se_male

We see that the sample mean BMI for women is expected to be off by around 0.14 relative to the population mean BMI for women, and the sample mean BMI for men is expected to be off by around 0.12 relative to the population mean BMI for men.

The standard error of the mean for women is slightly larger for women than for men. The reason for this is that even though the NHANES sample size for women is slightly larger than that for men, the data for women appears to be more spread out. The greater standard deviation for the female BMI values leads in turn to less precision when estimating the population mean BMI for females.

As was the case for proportions, the 95% confidence interval for the mean can be calculated by taking the estimate plus and minus 2 (or 1.96) times the standard error. The 95% confidence interval for female BMI is thus calculated as follows:

As was the case for proportions, the 95% confidence interval for the mean can be calculated by taking the estimate plus and minus 2 (or 1.96) times the standard error. The 95% confidence interval for female BMI is thus calculated as follows:

In [17]:
se = 1.96 * 7.753 / np.sqrt(2976)
lower_bound_fe = 29.94 - se
upper_bound_fe = 29.94 + se

lower_bound_fe, upper_bound_fe

(29.661446004723665, 30.218553995276338)

Below we show how the one-sample confidence interval can be calculated using Statsmodels. The numbers differ slightly due to rounding in the calculation above. The result below is more exact.

In [20]:
female_bmi = df.loc[df.RIAGENDR == 2, 'BMXBMI'].dropna()
sm.stats.DescrStatsW(female_bmi).zconfint_mean()

(29.659875498090155, 30.22001580625768)

### Confidence intervals for the difference between two means

Now we turn to studying the difference between two means, taking the difference between mean female and male BMI for illustration.  As discussed above, the standard error for the difference of two means taken from independent samples is sqrt(SE1^2 + SE2^2), where SE1 and SE2 are the standard errors for the two means being compared.  Below we see that this gives us a value around 0.19 when comparing the female BMI to the male BMI.  This is substantially larger than either the SEM for estimating the female mean (0.14) or the SEM for estimating the male mean (0.12).  It is expected that the standard error for the difference between two means is greater than the standard errors for estimating a single mean, since the uncertainty of both gender-specific proportions impacts the statistic.

In [21]:
se_diff = np.sqrt(se_female ** 2 + se_male ** 2)
se_diff

0.012729881381407434

We can can now construct a 95% confidence interval for the difference between the female and male mean BMI.

In [22]:
bmi_diff = 29.94 - 28.78
lower_bound = bmi_diff - 2 * se_diff
upper_bound = bmi_diff + 2 * se_diff

lower_bound, upper_bound

(1.1345402372371853, 1.185459762762815)

This finding indicates that while the point estimate shows that the women in our sample have around 1.1 unit greater BMI than the men in our sample, the true difference between the mean for all women in the population and for all men in the population could fall between 0.79 and 1.53, and still be consistent with the observed data.

Age-stratified confidence intervals As a final example, we refine the analysis above by considering the difference of mean BMI values between females and males within age bands. We see below that the overall average difference of 1.1 units results from differences that are very different based on age. Specifically, the difference between female and male BMI is much smaller than 1.1 for younger people, and much larger than 1.1 for older people.

Since the confidence bands for people under 40 contain 0, the data are consistent with there being no difference between female and male BMI in this age range. For people older than 40, a hypothetical zero difference between the mean BMI values for females and males is not very consistent with the data. Informally, we can say that the data strongly suggest that the female mean BMI is greater than the male mean BMI in this age band, with the difference being anywhere from 0.5 to 2 units.

In [25]:
pr = df.groupby(['agegrp', 'RIAGENDRx']).agg({'BMXBMI': ['mean', 'std', 'count']}).unstack()

BMXBMI                                           
                mean                  std            count     
RIAGENDRx     female       male    female      male female male
agegrp                                                         
(18, 30]   28.123881  27.391822  7.745893  6.649440    603  538
(30, 40]   30.325586  29.611726  8.315608  6.622412    469  452
(40, 50]   31.160643  29.724623  8.076195  6.407076    498  398
(50, 60]   30.743777  29.231486  7.575848  5.914373    466  451
(60, 70]   31.074828  29.392488  7.604514  5.933307    437  426
(70, 80]   29.138213  27.957692  6.284968  4.974855    403  390